neuraxle/checkpoints.py

```diff
@@ -19,185 +19,442 @@
+        :param context: context for creating the step checkpoint
+        :return: saved data container
+        :rtype: DataContainer
+        """
+        return self.save_checkpoint(data_container, context)
+
+    def handle_fit_transform(self, data_container: DataContainer, context: ExecutionContext) -> Tuple['Checkpoint', DataContainer]:
+        """
+        Saves step, and data checkpointers for the FIT_TRANSORM execution mode.
+
+        :param data_container: data container for creating the data checkpoint
+        :param context: context for creating the step checkpoint
+        :return: saved data container
+        :rtype: DataContainer
+        """
+        return self, self.save_checkpoint(data_container, context)
+
+    def save_checkpoint(self, data_container: DataContainer, context: ExecutionContext):
+        """
+        Saves step, and data checkpointers for the current execution mode.
+
+        :param data_container: data container for creating the data checkpoint
+        :param context: context for creating the step checkpoint
+        :return: saved data container
+        :rtype: DataContainer
+        """
+        for step_checkpointer in self.step_checkpointers:
+            if step_checkpointer.is_for_execution_mode(context.get_execution_mode()):
+                step_checkpointer.save_checkpoint(data_container, context)
+
+        for checkpointer in self.data_checkpointers:
+            if checkpointer.is_for_execution_mode(context.get_execution_mode()):
+                checkpointer.save_checkpoint(data_container, context)
+
+        return data_container
+
+    def read_checkpoint(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
+        """
+        Read data checkpoint for the current execution mode using self.data_checkpointers.
+
+        :param data_container: data container to load checkpoint from
+        :param context: execution mode to load checkpoint from
+        :return: loaded data container checkpoint
+        :rtype: DataContainer
+        """
+        for checkpointer in self.data_checkpointers:
+            if checkpointer.is_for_execution_mode(context.get_execution_mode()):
+                data_container = checkpointer.read_checkpoint(data_container, context)
+
+        return data_container
+
+    def should_resume(self, data_container: DataContainer, context: ExecutionContext) -> bool:
+        """
+        Returns True if all of the execution mode data checkpointers can be resumed.
+
+        :param context: execution context
+        :param data_container: data to resume
+        :return: if we can resume the checkpoint
+        :rtype: bool
+        """
+        for checkpointer in self.data_checkpointers:
+            if checkpointer.is_for_execution_mode(context.get_execution_mode()):
+                if not checkpointer.should_resume(data_container, context):
+                    return False
```

Comment:
I wonder what will happen when someone will mix data checkpoints without step checkpoints and vice versa, etc.

Reply:
Should we add exceptions or something ? I wasn't sure. 
Maybe people won't use one of the two dunno ? they could pass in not data checkpointers, or no step checkpointers

Reply:
Perhaps we could print warnings at the very least? I didn't really analyze and thought about all of that yet so I don't know. I'd like the framework to remain flexible, but yet safe to use.